docs/architecture-decision-records/0003-generalize-modules.md

```diff
@@ -0,0 +1,112 @@
+
+        // ApplyEvent applies a single input event to the module, making it advance its state
+        // and return a (potentially empty) list of output events that the application of the input event results in.
+        ApplyEvent(event *eventpb.Event) *events.EventList
+        
+        // Status returns the current state of the module.
+        // Mostly for debugging purposes.
+        Status() (s Status, err error) // The Status type is a placeholder not intended to be defined by this ADR.
+      }
+      ```
+    - `ActiveModule`:
+      Like the `PassiveModule`, the `ActiveModule` consumes input events. However,
+      unlike the `PassiveModule`, the `ActiveModule` does not synchronously return receive input
+      and return output in a function call.
+      Instead, the `ActiveModule` asynchronously reads input events from a channel
+      and writes produced events to an output channel.
+      In an `ActiveModule`, the produced events might or might not depend on the input events.
+      This makes the `ActiveModule` suitable for injecting external events
+      (e.g., reception of a message over the network) in the Node.
+      To prevent the module from flooding the Node with too many external events,
+      the node might ask it to temporarily back off and function essentially as a `PassiveModule`.
+      ```go
+      type ActiveModule interface {
+
+        // Run performs the module's processing.
+        // It usually reads events from eventsIn, processes them, and writes new events to eventsOut.
+        // Note, however, that the implementation may produce output events even without receiving any input.
+        // Before Run is called, ActiveModule guarantees not to read or write,
+        // respectively, from eventsIn and to eventsOut.
+        // Run must return after ctx is canceled.
+        Run(ctx context.Context, eventsIn <-chan *events.EventList, eventsOut chan<- *events.EventList) error
+
+        // BackOff asks the module to back off, i.e., to not produce new events
+        // except in immediate reaction to input events, until ctx is canceled.
+        // After BackOff is called and before ctx is canceled,
+        // the module will only produce a bounded number of output events and only in reaction to processing input events.
+        // In other words, if BackOff has been called, ctx has not been canceled, and the input channel is not written to,
+        // the module will eventually stop producing output events.
+        //
+        // This function is necessary for guaranteeing that the internal Node event buffers do not grow indefinitely.
+        // This could potentially happen when external events are injected into the system through active modules
+        // faster than the system can process them.
+        // In such a case, the Node asks all active modules to back off
+        // until enough events in the internal buffers have been processed.
+        BackOff(ctx context.Context)
```

Comment:
Assuming `Run` spawns a new goroutine to receive and produce events, a concurrent invocation to `BackOff` alone cannot strictly ensure a predictable boundary on the number of events produced until the invocation takes effect. To put an active module on hold, I suggest to simply stop receiving from `eventsOut` instead.

Reply:
> Technically, `Run` need not itself spawn a new goroutine, but I'd expect it to be called as a separate goroutine. But that's almost the same thing for many practical purposes, including this one.

It would be better to clarify in the comment if `Run` is supposed to be a blocking call.

> But we could still keep the `BackOff` function as a more graceful best-effort way of throttling the influx of external events when the event queues become big but not too big (some parameters can define what "big" and "too big" is). That way, internal events (e.g. timeouts) would still be processed by the active module and really only the external events (the module inherently knows which events it considers internal and which external) would be stopped, giving the node a chance to also process the "secondary" internal events (results of processing internal events) before the external ones are enabled again. If we do that, I'll reflect that in the comments ofc. Do you see any issues with that approach?

Not sure if we really need this level of control (and complexity), at least until we clearly realize the opposite. Do you have in mind a plausible example of a module that produces both external and internal events?